server/gitrest/packages/gitrest-base/src/utils/isomorphicgitManager.ts

```diff
@@ -359,78 +360,104 @@ export class IsomorphicGitRepositoryManager implements IRepositoryManager {
+                });
+            // services-client/getOrCreateRepository depends on a 400 response code
+            throw new NetworkError(400, `Repo does not exist ${args?.directoryPath}`);
+        };
 
-            this.repositoryCache.add(repoPath);
-        }
+        return this.repoManagerFactoryInternalHandler(params, onRepoNotExists, false);
+    }
+
+    public async openOrCreate(params: IRepoManagerParams): Promise<IsomorphicGitRepositoryManager> {
+        const onRepoNotExists = async () => {
+            await this.create(params);
+        };
 
-        const repoManager = new IsomorphicGitRepositoryManager(
-            fileSystemManager,
-            params.repoOwner,
-            params.repoName,
-            directoryPath,
-            lumberjackBaseProperties);
-        return repoManager;
+        return this.repoManagerFactoryInternalHandler(params, onRepoNotExists, false);
+    }
+
+    private async repoManagerFactoryInternalHandler(
+        params: IRepoManagerParams,
+        onRepoNotExists: (args?: any) => Promise<void> | void | never,
+        shouldUseMutex: boolean) {
+            const repoPath = helpers.getRepoPath(
+                this.repoPerDocEnabled,
+            params.repoName, /* tenantId */
+            params.storageRoutingId?.documentId,
+                this.storageDirectoryConfig.useRepoOwner ? params.repoOwner : undefined);
+            const directoryPath = helpers.getGitDirectory(
+                repoPath,
+                this.storageDirectoryConfig.baseDir);
+            const repoName =
+                this.repoPerDocEnabled ?
+                    `${params.repoName}/${params.storageRoutingId?.documentId}` :
+                    params.repoName;
+            const fileSystemManager = this.fileSystemManagerFactory.create(params.fileSystemManagerParams);
+            const lumberjackBaseProperties = helpers.getLumberjackBasePropertiesFromRepoManagerParams(params);
+
+            const handlerCore = async () => {
+                if (!(this.repositoryCache.has(repoPath))) {
+                    const repoExists = await helpers.exists(fileSystemManager, directoryPath);
+                    if (!repoExists || !repoExists.isDirectory()) {
+                        await onRepoNotExists({
+                            fileSystemManager,
+                            repoPath,
+                            directoryPath,
+                            lumberjackBaseProperties });
+                    } else {
+                        this.repositoryCache.add(repoPath);
+                    }
+                }
+
+                const repoManager = new IsomorphicGitRepositoryManager(
+                    fileSystemManager,
+                    params.repoOwner,
+                    repoName,
+                    directoryPath,
+                    lumberjackBaseProperties);
+                return repoManager;
+            };
+
+            if (shouldUseMutex) {
```

Comment:
>shouldUseMutex [](http://example.com/codeflow?start=16&length=14)

(educate me) Do I understand it correctly that we instantiate IsomorphicGitRepositoryManager for every GitRest call? if so, we wouldn't need a mutex, right?

Reply:
Ha...you explained the race condition in the one-pager :smile: